man/queue/sp_queue_set.3
.\"  Staple - A general-purpose data structure library in pure C89.
.\"  Copyright (C) 2021  Randoragon
.\"
.\"  This library is free software; you can redistribute it and/or
.\"  modify it under the terms of the GNU Lesser General Public
.\"  License as published by the Free Software Foundation;
.\"  version 2.1 of the License.
.\"
.\"  This library is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\"  Lesser General Public License for more details.
.\"
.\"  You should have received a copy of the GNU Lesser General Public
.\"  License along with this library; if not, write to the Free Software
.\"  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
.\"--------------------------------------------------------------------------------
.TH SP_QUEUE_SET 3 DATE "libstaple-VERSION"
.SH NAME
sp_queue_set,
sp_queue_setc,
sp_queue_sets,
sp_queue_seti,
sp_queue_setl,
sp_queue_setsc,
sp_queue_setuc,
sp_queue_setus,
sp_queue_setui,
sp_queue_setul,
sp_queue_setf,
sp_queue_setd,
sp_queue_setld,
sp_queue_setll,
sp_queue_setull,
sp_queue_setu8,
sp_queue_setu16,
sp_queue_setu32,
sp_queue_setu64,
sp_queue_seti8,
sp_queue_seti16,
sp_queue_seti32,
sp_queue_seti64,
sp_queue_setstr,
sp_queue_setstrn
\- set the value of an arbitrary element in a queue
.SH SYNOPSIS
.ad l
#include <staple.h>
.sp
int
.BR sp_queue_set "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
void
.RI * val )
.br
int
.BR sp_queue_setc "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
char
.IR val )
.br
int
.BR sp_queue_sets "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
short
.IR val )
.br
int
.BR sp_queue_seti "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
int
.IR val )
.br
int
.BR sp_queue_setl "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
long
.IR val )
.br
int
.BR sp_queue_setsc "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
signed char
.IR val )
.br
int
.BR sp_queue_setuc "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
unsigned char
.IR val )
.br
int
.BR sp_queue_setus "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
unsigned short
.IR val )
.br
int
.BR sp_queue_setui "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
unsigned int
.IR val )
.br
int
.BR sp_queue_setul "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
unsigned long
.IR val )
.br
int
.BR sp_queue_setf "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
float
.IR val )
.br
int
.BR sp_queue_setd "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
double
.IR val )
.br
int
.BR sp_queue_setld "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
long double
.IR val )
.br
int
.BR sp_queue_setll "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
long long
.IR val )
.br
int
.BR sp_queue_setull "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
unsigned long long
.IR val )
.br
int
.BR sp_queue_setu8 "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
uint8_t
.IR val )
.br
int
.BR sp_queue_setu16 "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
uint16_t
.IR val )
.br
int
.BR sp_queue_setu32 "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
uint32_t
.IR val )
.br
int
.BR sp_queue_setu64 "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
uint64_t
.IR val )
.br
int
.BR sp_queue_seti8 "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
int8_t
.IR val )
.br
int
.BR sp_queue_seti16 "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
int16_t
.IR val )
.br
int
.BR sp_queue_seti32 "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
int32_t
.IR val )
.br
int
.BR sp_queue_seti64 "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
int64_t
.IR val )
.br
int
.BR sp_queue_setstr "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
const char
.RI * val )
.br
int
.BR sp_queue_setstrn "(const struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
const char
.RI * val ,
size_t
.IR len )
.sp
Link with \fI-lstaple\fP.
.ad
.SH DESCRIPTION
.P
The functions in the
.BR sp_queue_set ()
family are used to set the value of an arbitrary element in a queue. The
previous value is overwritten and non-recoverable.
.P
.I queue
is the queue.
.P
.I idx
is the index of the element that shall be overwritten. Indices start from 0 and
extend to the rear \(em 0 is the front element, 1 is the element behind it, and
so on.
.br
.I idx
must be smaller or equal to
.IR queue->size .
.P
.I val
is the new value for the element at position
.IR idx .
.SS Generic Form
.I val
must be a void pointer to the first byte of the new value. The source and
destination blocks must not overlap in memory. The number of bytes to copy is
equal to
.IR queue->elem_size
field.
.SS Suffixed Form
.I val
is passed directly in the function argument.
.P
The
.BR sp_queue_setstr ()
function creates a copy of the string pointed to by
.I val
and stores the new pointer on the queue. The string must fit in
.I SIZE_MAX
bytes, including the null terminator. The previous string is freed.
.P
The
.BR sp_queue_setstrn ()
function is equivalent to
.BR sp_queue_setstr (),
except it shall always copy exactly
.I len
bytes and append a null terminator at the end.
.I len
must be smaller than
.IR SIZE_MAX .
.SH RETURN VALUE
If successful, the functions in the
.BR sp_queue_set ()
family shall return 0. Otherwise, they shall return a code indicating the
error.
.SH ERRORS
The functions in the
.BR sp_queue_set ()
family shall fail if:
.IP \fBSP_EINVAL\fP 1.5i
.I queue
is a null pointer (debug mode only).
.IP \fBSP_EINDEX\fP 1.5i
.I idx
out of range (debug mode only).
.SS Generic Form
.IP \fBSP_EINVAL\fP 1.5i
.I val
is a null pointer (debug mode only).
.SS Suffixed Form
.IP \fBSP_EILLEGAL\fP 1.5i
.IR queue->elem_size
does not match the size of the
.I val
argument (debug mode only).
.P
The
.BR sp_queue_setstr ()
and
.BR sp_queue_setstrn ()
functions shall fail if:
.IP \fBSP_EINVAL\fP 1.5i
.I val
is a null pointer (debug mode only).
.IP \fBSP_ERANGE\fP 1.5i
string too long or
.I len
too large.
.IP \fBSP_ENOMEM\fP 1.5i
Memory allocation failed.
.SH CONFORMING TO
See
.BR libstaple (7).
.SH SEE ALSO
.ad l
.BR libstaple (7),
.BR sp_queue (7),
.BR sp_queue_create (3),
.BR sp_queue_destroy (3),
.BR sp_queue_clear (3),
.BR sp_queue_push (3),
.BR sp_queue_peek (3),
.BR sp_queue_pop (3),
.BR sp_queue_insert (3),
.BR sp_queue_remove (3),
.BR sp_queue_qinsert (3),
.BR sp_queue_qremove (3),
.BR sp_queue_get (3),
.BR sp_queue_eq (3),
.BR sp_queue_copy (3),
.BR sp_queue_map (3),
.BR sp_queue_print (3)


src/queue/sp_queue_set.c
/*  Staple - A general-purpose data structure library in pure C89.
 *  Copyright (C) 2021  Randoragon
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation;
 *  version 2.1 of the License.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "../sp_queue.h"
#include "../internal.h"
#include "../sp_errcodes.h"
#include <string.h>

int sp_queue_set(struct sp_queue *queue, size_t idx, void *val)
{
	char *p;
#ifdef STAPLE_DEBUG
	if (queue == NULL) {
		error(("queue is NULL"));
		return SP_EINVAL;
	}
	if (val == NULL) {
		error(("val is NULL"));
		return SP_EINVAL;
	}
	if (idx >= queue->size) {
		error(("index out of range"));
		return SP_EINDEX;
	}
#endif
	p = sp_ringbuf_get(idx, queue->data, queue->capacity, queue->elem_size, queue->head);
	memcpy(p, val, queue->elem_size);
	return 0;
}
